commit 2fa277df78416d75ab540ab64168e356d05ba5cc
Author: Stefan Saraev <stefan@saraev.ca>
Date:   Tue Jul 23 11:31:22 2013 +0300

    do not touch nat table before enabling tethering
    
    TODO: remove when upstream does a better fix

diff --git a/src/connman.h b/src/connman.h
index bc7a50d..e280bd7 100644
--- a/src/connman.h
+++ b/src/connman.h
@@ -913,6 +913,8 @@ int __connman_nat_enable(const char *name, const char *address,
 				unsigned char prefixlen);
 void __connman_nat_disable(const char *name);
 
+static bool nat_enabled = false;
+
 struct firewall_context;
 
 struct firewall_context *__connman_firewall_create(void);
diff --git a/src/firewall.c b/src/firewall.c
index bf05294..8ab34d9 100644
--- a/src/firewall.c
+++ b/src/firewall.c
@@ -421,7 +421,8 @@ static void flush_all_tables(void)
 
 	flush_table("filter");
 	flush_table("mangle");
-	flush_table("nat");
+	if (nat_enabled)
+		flush_table("nat");
 }
 
 int __connman_firewall_init(void)
diff --git a/src/nat.c b/src/nat.c
index 8d307eb..11c9def 100644
--- a/src/nat.c
+++ b/src/nat.c
@@ -81,6 +81,8 @@ static int enable_nat(struct connman_nat *nat)
 	if (err < 0)
 		return err;
 
+	nat_enabled = true;
+
 	return __connman_firewall_enable(nat->fw);
 }
 
